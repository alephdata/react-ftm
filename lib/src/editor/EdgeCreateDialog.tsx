import * as React from 'react'
import { Dialog, Menu, MenuItem, FormGroup, Intent, Button, ControlGroup, InputGroup } from '@blueprintjs/core'
import { Model, Schema, Property, IconRegistry } from '@alephdata/followthemoney'
import { GraphContext, IGraphContext } from '../GraphContext'


interface IEdgeCreateDialogProps {
  isOpen: boolean,
  toggleDialog: () => any
}

interface IEdgeCreateDialogState {
  label: string,
  schema?: Schema
}

export class EdgeType {
  public schema?: Schema
  public property?: Property

  constructor(schema?: Schema, property?: Property) {
    this.schema = schema
    this.property = property
  }

  static getAll(model: Model): EdgeType[] {
    const types = new Array<EdgeType>()
    if (model.schemata) {
      model.schemata.forEach((schema) => {
        types.push(new EdgeType(schema))
      })
    }
    
    return types
  }
}


export class EdgeCreateDialog extends React.Component<IEdgeCreateDialogProps, IEdgeCreateDialogState> {
  static contextType = GraphContext;
  context!: React.ContextType<typeof GraphContext>;
  state: IEdgeCreateDialogState = {
    label: ''
  }

  constructor(props: any) {
    super(props);
    // this.onChangeLabel = this.onChangeLabel.bind(this);
    // this.onSchemaSelect = this.onSchemaSelect.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  // onChangeLabel(e: React.ChangeEvent<HTMLInputElement>) {
  //   this.setState({ label: e.target.value })
  // }

  // getSchema(): Schema {
  //   const { layout } = this.context as IGraphContext
  //   return this.state.schema || layout.model.getSchema('Person')
  // }

  // onSchemaSelect(schema: Schema) {
  //   this.setState({ schema })
  // }

  onSubmit(e: React.ChangeEvent<HTMLFormElement>) {
    // const { label } = this.state
    // const schema = this.getSchema()
    // const { layout, updateLayout } = this.context as IGraphContext
    e.preventDefault()
    // if (this.checkValid()) {
    //   const entity = layout.model.createEntity(schema)
    //   entity.setProperty('name', label)
    //   layout.addEntity(entity)
    //   layout.layout()
    //   const vertex = layout.getVertexByEntity(entity)
    //   if (vertex) {
    //     layout.selectVertex(vertex)
    //     layout.viewport = layout.viewport.setCenter(vertex.position)
    //     updateLayout(layout)
    //     this.setState({label: ''})
    //     this.props.toggleDialog()
    //   }
    // }
  }

  // checkValid(): boolean {
  //   const { label } = this.state;
  //   if (label.trim().length < 3) {
  //     return false;
  //   }
  //   return true;
  // }

  render() {
    // const { layout } = this.context as IGraphContext
    const { isOpen, toggleDialog } = this.props
    // const schema = this.getSchema()
    // const placeholder = `${schema.label} name`
    // const isValid = this.checkValid()
    return (
      <Dialog icon="new-link" isOpen={isOpen} title="Add connection" onClose={toggleDialog}>
        <form onSubmit={this.onSubmit}>
          {/* <div className="bp3-dialog-body">
            <ControlGroup fill>
              <VertexSchemaSelect
                model={layout.model}
                schema={schema}
                onSelect={this.onSchemaSelect}
              />
              <InputGroup
                large
                intent={isValid ? undefined : Intent.WARNING}
                className="bp3-fill"
                value={this.state.label}
                onChange={this.onChangeLabel}
                placeholder={placeholder}
              />
            </ControlGroup>
          </div> */}
        </form>
      </Dialog>
    );
  }
}
