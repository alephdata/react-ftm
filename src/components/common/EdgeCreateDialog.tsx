import * as React from 'react'
import { compose } from 'redux';
import { connect, ConnectedProps } from 'react-redux';
import { defineMessages, injectIntl, WrappedComponentProps } from 'react-intl';
import { FormGroup, Intent, Button } from '@blueprintjs/core'
import { Entity, Model, Schema, Values } from '@alephdata/followthemoney'

import { IEntityContext } from 'contexts/EntityContext';
import { EdgeTypeSelect, EntitySuggest } from 'editors'
import { EdgeType } from 'types'
import { Dialog } from 'components/common'

const messages = defineMessages({
  add_link: {
    id: 'dialog.edge_create.title',
    defaultMessage: 'Add link',
  },
  source: {
    id: 'dialog.edge_create.source_label',
    defaultMessage: 'Source',
  },
  target: {
    id: 'dialog.edge_create.target_label',
    defaultMessage: 'Target',
  },
  type: {
    id: 'dialog.edge_create.type_label',
    defaultMessage: 'Type',
  },
  type_select: {
    id: 'dialog.edge_create.type_placeholder',
    defaultMessage: 'Select type',
  },
  submit: {
    id: 'dialog.edge_create.submit',
    defaultMessage: 'Create',
  },
});

interface IEdgeCreateDialogProps extends WrappedComponentProps {
  source: Entity
  target?: Entity
  isOpen: boolean,
  toggleDialog: () => any
  onSubmit: (source: Entity, target: Entity, type: EdgeType) => void
  entityContext: IEntityContext
}

interface IEdgeCreateDialogState {
  sourceQueryText: string
  targetQueryText: string
  source?: Entity
  target?: Entity
  type?: EdgeType
  isProcessing: boolean
}

class EdgeCreateDialogBase extends React.Component<IEdgeCreateDialogProps & PropsFromRedux, IEdgeCreateDialogState> {
  types: EdgeType[] = []
  state: IEdgeCreateDialogState = {
    isProcessing: false,
    sourceQueryText: '',
    targetQueryText: ''
  }

  constructor(props: IEdgeCreateDialogProps & PropsFromRedux) {
    super(props)

    this.types = EdgeType.getAll(props.model)

    this.onSelectSource = this.onSelectSource.bind(this)
    this.onSelectTarget = this.onSelectTarget.bind(this)
    this.onChangeType = this.onChangeType.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onReverse = this.onReverse.bind(this)
  }

  componentDidUpdate(prevProps: IEdgeCreateDialogProps) {
    const { isOpen, source, target } = this.props;
    if (!prevProps.isOpen && isOpen) {
      this.setState({ source, target, type: undefined });
    }
  }

  onChangeType(type: EdgeType) {
    const { source, target } = this.state
    if (source && target) {
      if (!type.match(source, target) && type.match(target, source)) {
        this.setState({ source: target, target: source })
      }
      this.setState({ type })
    }
  }

  onSelectSource(source: Entity) {
    this.setState(({ target }) => ({
      source,
      target: target?.id === source.id ? undefined : target,
      type: undefined
    }))
  }

  onSelectTarget(target: Entity) {
    this.setState(({ source }) => ({
      target,
      source: source?.id === target.id ? undefined : source,
      type: undefined
    }))
  }

  isValid() {
    const { source, target, type } = this.state
    return source && target && type && type.match(source, target)
  }

  async onSubmit(e: React.ChangeEvent<HTMLFormElement>) {
    const { onSubmit, toggleDialog } = this.props
    const { source, target, type } = this.state
    e.preventDefault()
    if (source && target && type && this.isValid()) {
      onSubmit(source, target, type);
      toggleDialog();
    }
  }

  isReversible() {
    const { source, target, type } = this.state
    return source && target && type && type.match(target, source)
  }

  onReverse() {
    const { source, target } = this.state
    if (this.isReversible()) {
      this.setState({ source: target, target: source })
    }
  }

  getTypes(): EdgeType[] {
    const { source, target } = this.state

    if (source && target) {
      return this.types.filter((et) => et.match(source, target) || et.match(target, source))
    }
    return []
  }

  getSourceLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.source).label
      }
      if (type.property) {
        return type.property.schema.label
      }
    }
  }

  getTargetLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.target).label
      }
      if (type.property) {
        return type.property.label
      }
    }
  }

  getTypeDescription(): string | undefined | null {
    const { type } = this.state
    if (type) {
      if (type.schema) {
        return type.schema.description
      }
      if (type.property) {
        return type.property.description
      }
    }
  }

  render() {
    const { entityContext, intl, isOpen, suggestionSchemata, toggleDialog } = this.props
    const { isProcessing, source, target, type, sourceQueryText, targetQueryText } = this.state
    const types = this.getTypes()

    return (
      <Dialog
        icon="new-link"
        isOpen={isOpen}
        isProcessing={isProcessing}
        title={intl.formatMessage(messages.add_link)}
        onClose={() => toggleDialog()}
        className="large"
      >
        <form onSubmit={this.onSubmit}>
          <div className="bp3-dialog-body">
            <div style={{flex: 1, display: 'flex', flexFlow: 'row'}}>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.source)} helperText={this.getSourceLabel()}>
                  <EntitySuggest
                    onSubmit={(selected: Array<Entity>) => this.onSelectSource(selected?.[0])}
                    onQueryChange={(query: string) => this.setState({ sourceQueryText: query })}
                    queryText={sourceQueryText}
                    schemata={suggestionSchemata}
                    entityContext={entityContext}
                    suggestLocalEntities={true}
                    entitySelectProps={{
                      values: source ? [source] : [],
                      allowMultiple: false
                    }}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.type)} helperText={this.getTypeDescription()}>
                  <EdgeTypeSelect
                    items={types}
                    value={type}
                    onChange={this.onChangeType}
                    placeholder={intl.formatMessage(messages.type_select)}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.target)} helperText={this.getTargetLabel()}>
                  <EntitySuggest
                    onSubmit={(selected: Array<Entity>) => this.onSelectTarget(selected?.[0])}
                    onQueryChange={(query: string) => this.setState({ targetQueryText: query })}
                    queryText={targetQueryText}
                    schemata={suggestionSchemata}
                    entityContext={entityContext}
                    suggestLocalEntities={true}
                    entitySelectProps={{
                      values: target ? [target] : [],
                      allowMultiple: false
                    }}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 0, flexShrink: 1, flexBasis: '1%'}}>
                <FormGroup label='&nbsp;'>
                  <Button
                    onClick={this.onReverse}
                    disabled={!this.isReversible()}
                    icon="exchange"
                  />
                </FormGroup>
              </div>
            </div>
          </div>
          <div className="bp3-dialog-footer">
            <div className="bp3-dialog-footer-actions">
              <Button
                intent={Intent.PRIMARY}
                disabled={!this.isValid()}
                text={intl.formatMessage(messages.submit)}
                type="submit"
              />
            </div>
          </div>
        </form>
      </Dialog>
    );
  }
}

const mapStateToProps = (state: any, ownProps: IEdgeCreateDialogProps) => {
  const { entityContext } = ownProps;
  const model = entityContext.selectModel(state);
  return ({
    model,
    suggestionSchemata: model.getSchemata()
      .filter((schema: Schema) => schema.isThing() && !schema.generated && !schema.abstract)
  });
}

const connector = connect(mapStateToProps);
type PropsFromRedux = ConnectedProps<typeof connector>

export const EdgeCreateDialog = connector(
  injectIntl(EdgeCreateDialogBase)
);
