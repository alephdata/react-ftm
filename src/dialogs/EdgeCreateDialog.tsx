import * as React from 'react'
import { defineMessages, WrappedComponentProps } from 'react-intl';
import _ from 'lodash';
import { Menu, MenuItem, FormGroup, Icon, Intent, Button, Alignment, Position } from '@blueprintjs/core'
import { Select, IItemListRendererProps, IItemRendererProps } from '@blueprintjs/select';
import { Entity } from '@alephdata/followthemoney'

import { EntitySelect } from '../editors'
import { EdgeType } from '../components'
import { IGraphContext } from '../GraphContext'
import { EntityManager } from '../EntityManager'
import { Schema } from '../types';

import Dialog from './Dialog';

const messages = defineMessages({
  add_link: {
    id: 'dialog.edge_create.title',
    defaultMessage: 'Add link',
  },
  source: {
    id: 'dialog.edge_create.source_label',
    defaultMessage: 'Source',
  },
  target: {
    id: 'dialog.edge_create.target_label',
    defaultMessage: 'Target',
  },
  type: {
    id: 'dialog.edge_create.type_label',
    defaultMessage: 'Type',
  },
  type_select: {
    id: 'dialog.edge_create.type_placeholder',
    defaultMessage: 'Select type',
  },
  submit: {
    id: 'dialog.edge_create.submit',
    defaultMessage: 'Create',
  },
});

const EdgeTypeSelect = Select.ofType<EdgeType>();

interface IEdgeCreateDialogProps extends WrappedComponentProps {
  source: Entity
  target?: Entity
  isOpen: boolean,
  toggleDialog: () => any
  entityManager: EntityManager
}

interface IEdgeCreateDialogState {
  source?: Entity
  target?: Entity
  sourceSuggestions: any
  targetSuggestions: any
  type?: EdgeType
  isProcessing: boolean
}

export class EdgeCreateDialog extends React.Component<IEdgeCreateDialogProps, IEdgeCreateDialogState> {
  types: EdgeType[] = []
  state: IEdgeCreateDialogState = {
    isProcessing: false,
    sourceSuggestions: { isPending: false, results: [] },
    targetSuggestions: { isPending: false, results: [] },
  }

  constructor(props: any) {
    super(props)
    this.onSelectSource = this.onSelectSource.bind(this)
    this.onSelectTarget = this.onSelectTarget.bind(this)
    this.onChangeType = this.onChangeType.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onReverse = this.onReverse.bind(this)
    this.renderEdgeType = this.renderEdgeType.bind(this)
  }


  componentDidMount() {
    const { entityManager, source, target } = this.props
    this.setState({ source, target })
    this.types = EdgeType.getAll(entityManager.model)
  }

  componentDidUpdate(prevProps: IEdgeCreateDialogProps) {
    if (prevProps.source !== this.props.source || prevProps.target !== this.props.target) {
      const { source, target } = this.props
      this.setState({ source, target, type: undefined })
    }
  }

  onChangeType(type: EdgeType) {
    const { source, target } = this.state
    if (source && target) {
      if (!type.match(source, target) && type.match(target, source)) {
        this.setState({ source: target, target: source })
      }
      this.setState({ type })
    }
  }

  onSelectSource(source: Entity) {
    this.setState({ source, type: undefined })
  }

  onSelectTarget(target: Entity) {
    this.setState({ target, type: undefined })
  }

  isValid() {
    const { source, target, type } = this.state
    return source && target && type && type.match(source, target)
  }

  async onSubmit(e: React.ChangeEvent<HTMLFormElement>) {
    const { onSubmit } = this.props
    const { source, target, type } = this.state
    e.preventDefault()
    if (source && target && type && this.isValid()) {
      onSubmit({ source, target, type});
    }
  }

  isReversible() {
    const { source, target, type } = this.state
    return source && target && type && type.match(target, source)
  }

  onReverse() {
    const { source, target, type } = this.state
    if (this.isReversible()) {
      this.setState({ source: target, target: source })
    }
  }

  getTypes(): EdgeType[] {
    const { source, target } = this.state

    if (source && target) {
      return this.types.filter((et) => et.match(source, target) || et.match(target, source))
    }
    return []
  }

  getSourceLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.source).label
      }
      if (type.property) {
        return type.property.schema.label
      }
    }
  }

  getTargetLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.target).label
      }
      if (type.property) {
        return type.property.label
      }
    }
  }

  getTypeDescription(): string | undefined | null {
    const { type } = this.state
    if (type) {
      if (type.schema) {
        return type.schema.description
      }
      if (type.property) {
        return type.property.description
      }
    }
  }

  getEdgeTypeIcon(type?: EdgeType) {
    if (type?.schema) {
      return <Schema.Icon schema={type.schema} />
    } else {
      return <Icon icon="link" />
    }
  }

  renderEdgeTypeList(props: IItemListRendererProps<EdgeType>) {
    const { items, itemsParentRef, renderItem } = props;
    const [propertyEdgeTypes, entityEdgeTypes] = _.partition(
      items,
      (et: EdgeType) => et.isPropertyEdgeType()
    );
    return (
      <Menu ulRef={itemsParentRef}>
        {entityEdgeTypes.map(renderItem)}
        <Menu.Divider />
        {propertyEdgeTypes.map(renderItem)}
      </Menu>
    );
  }

  renderEdgeType(type: EdgeType, { handleClick, modifiers }:IItemRendererProps) {
    return <MenuItem
      active={modifiers.active}
      key={type.key}
      text={type.label}
      icon={this.getEdgeTypeIcon(type)}
      onClick={handleClick}
    />
  }

  async fetchEntitySuggestions(query: string, which?: string) {
    const stateKey = `${which}Suggestions`;

    console.log('fetching', query, which);

    const { entityManager } = this.props;
    this.setState({ [stateKey]: { isProcessing: true, results: [] } });
    const results = await entityManager.getEntitySuggestions(query, ['Thing']);
    console.log('results', which, results);
    this.setState({ [stateKey]: { isProcessing: false, results } });
  }

  render() {
    const { intl, isOpen, toggleDialog } = this.props
    const { isProcessing, source, target, type, sourceSuggestions, targetSuggestions } = this.state
    const types = this.getTypes()

    return (
      <Dialog
        icon="new-link"
        isOpen={isOpen}
        isProcessing={isProcessing}
        title={intl.formatMessage(messages.add_link)}
        onClose={() => toggleDialog()}
        className="large"
      >
        <form onSubmit={this.onSubmit}>
          <div className="bp3-dialog-body">
            <div style={{flex: 1, display: 'flex', flexFlow: 'row'}}>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.source)} helperText={this.getSourceLabel()}>
                  <EntitySelect
                    onSubmit={(selected: Array<Entity>) => this.onSelectSource(selected?.[0])}
                    onChange={(selected: Array<Entity>) => this.onSelectSource(selected?.[0])}
                    values={source ? [source] : []}
                    allowMultiple={false}
                    isFetching={sourceSuggestions.isPending}
                    entitySuggestions={sourceSuggestions.results}
                    onQueryChange={(query) => this.fetchEntitySuggestions(query, 'source')}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.type)} helperText={this.getTypeDescription()}>
                  <EdgeTypeSelect
                    popoverProps={{
                      position: Position.BOTTOM_LEFT,
                      minimal: true,
                      targetProps: {style: {width: '100%'}}
                    }}
                    filterable={false}
                    items={types}
                    itemListRenderer={this.renderEdgeTypeList}
                    itemRenderer={this.renderEdgeType}
                    onItemSelect={this.onChangeType}
                  >
                    <Button fill
                      disabled={!types.length}
                      text={type ? type.label : intl.formatMessage(messages.type_select)}
                      alignText={Alignment.LEFT}
                      icon={this.getEdgeTypeIcon(type)}
                      rightIcon='double-caret-vertical'
                    />
                  </EdgeTypeSelect>
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label={intl.formatMessage(messages.target)} helperText={this.getTargetLabel()}>
                  <EntitySelect
                    onSubmit={(selected: Array<Entity>) => this.onSelectTarget(selected?.[0])}
                    onChange={(selected: Array<Entity>) => this.onSelectTarget(selected?.[0])}
                    values={target ? [target] : []}
                    allowMultiple={false}
                    isFetching={targetSuggestions.isPending}
                    entitySuggestions={targetSuggestions.results}
                    onQueryChange={(query) => this.fetchEntitySuggestions(query, 'target')}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 0, flexShrink: 1, flexBasis: '1%'}}>
                <FormGroup label='&nbsp;'>
                  <Button
                    onClick={this.onReverse}
                    disabled={!this.isReversible()}
                    icon="exchange"
                  />
                </FormGroup>
              </div>
            </div>
          </div>
          <div className="bp3-dialog-footer">
            <div className="bp3-dialog-footer-actions">
              <Button
                intent={Intent.PRIMARY}
                disabled={!this.isValid()}
                text={intl.formatMessage(messages.submit)}
                type="submit"
              />
            </div>
          </div>
        </form>
      </Dialog>
    );
  }
}
