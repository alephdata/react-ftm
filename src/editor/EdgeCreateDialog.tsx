import * as React from 'react'
import { Dialog, MenuItem, FormGroup, Intent, Button, Alignment, Position } from '@blueprintjs/core'
import { Select, IItemRendererProps } from '@blueprintjs/select';
import { IGraphContext } from '../GraphContext'
import { VertexSelect } from './VertexSelect'
import { EdgeType } from './EdgeType'
import { Vertex,Edge } from '../layout';
import { SchemaIcon } from '../types';

const EdgeTypeSelect = Select.ofType<EdgeType>();

interface IEdgeCreateDialogProps extends IGraphContext {
  source: Vertex
  target?: Vertex
  isOpen: boolean,
  toggleDialog: () => any
}

interface IEdgeCreateDialogState {
  source?: Vertex
  target?: Vertex
  type?: EdgeType
}

export class EdgeCreateDialog extends React.Component<IEdgeCreateDialogProps, IEdgeCreateDialogState> {
  types: EdgeType[] = []
  state: IEdgeCreateDialogState = {}

  constructor(props: any) {
    super(props)
    this.onSelectSource = this.onSelectSource.bind(this)
    this.onSelectTarget = this.onSelectTarget.bind(this)
    this.onChangeType = this.onChangeType.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onReverse = this.onReverse.bind(this)
  }

  static getEdgeTypeIcon(type?: EdgeType) {
    if (type && type.schema) {
      return SchemaIcon.get(type.schema)
    }
    return 'link'
  }

  componentDidMount() {
    const { layout, source, target } = this.props
    this.setState({ source, target })
    this.types = EdgeType.getAll(layout.model)
  }

  componentDidUpdate(prevProps: IEdgeCreateDialogProps) {
    if (prevProps.source !== this.props.source || prevProps.target !== this.props.target) {
      const { source, target } = this.props
      this.setState({ source, target, type: undefined })
    }
  }

  onChangeType(type: EdgeType) {
    const { source, target } = this.state
    if (source && target) {
      if (!type.match(source, target) && type.match(target, source)) {
        this.setState({ source: target, target: source })
      }
      this.setState({ type })
    }
  }

  onSelectSource(source: Vertex) {
    this.setState({ source, type: undefined })
  }

  onSelectTarget(target: Vertex) {
    this.setState({ target, type: undefined })
  }

  isValid() {
    const { source, target, type } = this.state
    return source && target && type && type.match(source, target)
  }

  onSubmit(e: React.ChangeEvent<HTMLFormElement>) {
    const { layout, viewport, updateLayout, updateViewport, toggleDialog } = this.props
    const { source, target, type } = this.state
    e.preventDefault()
    if (source && target && type && this.isValid()) {
      const sourceEntity = source.getEntity()
      const targetEntity = target.getEntity()
      if (type.property && sourceEntity) {
        const value = targetEntity || target.label
        sourceEntity.setProperty(type.property, value)
        layout.addEntity(sourceEntity)
        const edge = Edge.fromValue(layout, type.property, source, target)
        layout.selectElement(edge)
        updateViewport(viewport.setCenter(edge.getCenter()), {animate:true})
      }
      if (type.schema && type.schema.edge && sourceEntity && targetEntity) {
        const entity = layout.model.createEntity(type.schema)
        entity.setProperty(type.schema.edge.source, sourceEntity)
        entity.setProperty(type.schema.edge.target, targetEntity)
        layout.addEntity(entity)
        const edge = Edge.fromEntity(layout, entity, source, target)
        layout.selectElement(edge)
        updateViewport(viewport.setCenter(edge.getCenter()), {animate:true})
      }
      updateLayout(layout, {modifyHistory:true})
      toggleDialog()
    }
  }

  isReversible() {
    const { source, target, type } = this.state
    return source && target && type && type.match(target, source)
  }

  onReverse() {
    const { source, target, type } = this.state
    if (this.isReversible()) {
      this.setState({ source: target, target: source })
    }
  }

  getTypes(): EdgeType[] {
    const { source, target } = this.state
    if (source && target) {
      return this.types.filter((et) => et.match(source, target) || et.match(target, source))
    }
    return []
  }

  getVertices(include?: Vertex, except?: Vertex): Vertex[] {
    const { layout } = this.props
    return layout.getVertices()
      .filter((vertex) => {
        const isInclude = include ? include.id === vertex.id : false
        const isExcept = except ? except.id === vertex.id : false
        return isInclude || !(isExcept || vertex.isHidden())
      })
      .sort((a, b) => a.label.localeCompare(b.label))
  }

  getSourceLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.source).label
      }
      if (type.property) {
        return type.property.schema.label
      }
    }
  }

  getTargetLabel(): string | undefined {
    const { type } = this.state
    if (type) {
      if (type.schema && type.schema.edge) {
        return type.schema.getProperty(type.schema.edge.target).label
      }
      if (type.property) {
        return type.property.label
      }
    }
  }

  getTypeDescription(): string | undefined | null {
    const { type } = this.state
    if (type) {
      if (type.schema) {
        return type.schema.description
      }
      if (type.property) {
        return type.property.description
      }
    }
  }

  renderEdgeType(type: EdgeType, { handleClick, modifiers }: IItemRendererProps) {
    if (!modifiers.matchesPredicate) {
        return null;
    }
    return <MenuItem
      active={modifiers.active}
      key={type.key}
      text={type.label}
      icon={EdgeCreateDialog.getEdgeTypeIcon(type)}
      onClick={handleClick}
    />
  }

  render() {
    const { isOpen, toggleDialog } = this.props
    const { source, target, type } = this.state
    const types = this.getTypes()

    console.log('source and target', source, target);
    return (
      <Dialog icon="new-link" isOpen={isOpen} title="Add link" onClose={toggleDialog} style={{width: '800px'}}>
        <form onSubmit={this.onSubmit}>
          <div className="bp3-dialog-body">
            <div style={{flex: 1, display: 'flex', flexFlow: 'row'}}>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label='Source' helperText={this.getSourceLabel()}>
                  <VertexSelect
                    vertices={this.getVertices(source, target)}
                    vertex={source}
                    onSelect={this.onSelectSource}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label="Type" helperText={this.getTypeDescription()}>
                  <EdgeTypeSelect
                    popoverProps={{
                      position: Position.BOTTOM_LEFT,
                      minimal: true,
                      targetProps: {style: {width: '100%'}}
                    }}
                    filterable={false}
                    items={types}
                    itemRenderer={this.renderEdgeType}
                    onItemSelect={this.onChangeType}
                  >
                    <Button fill
                      disabled={!types.length}
                      text={type ? type.label : 'Select type'}
                      alignText={Alignment.LEFT}
                      icon={EdgeCreateDialog.getEdgeTypeIcon(type)}
                      rightIcon='double-caret-vertical'
                    />
                  </EdgeTypeSelect>
                </FormGroup>
              </div>
              <div style={{flexGrow: 1, flexShrink: 1, flexBasis: 'auto', paddingRight: '1em'}}>
                <FormGroup label='Target' helperText={this.getTargetLabel()}>
                  <VertexSelect
                    vertices={this.getVertices(target, source)}
                    vertex={target}
                    onSelect={this.onSelectTarget}
                  />
                </FormGroup>
              </div>
              <div style={{flexGrow: 0, flexShrink: 1, flexBasis: '1%'}}>
                <FormGroup label='&nbsp;'>
                  <Button
                    onClick={this.onReverse}
                    disabled={!this.isReversible()}
                    icon="exchange"
                  />
                </FormGroup>
              </div>
            </div>
          </div>
          <div className="bp3-dialog-footer">
          <div className="bp3-dialog-footer-actions">
            <Button
              intent={Intent.PRIMARY}
              disabled={!this.isValid()}
              text="Create"
              type="submit"
            />
          </div>
        </div>
        </form>
      </Dialog>
    );
  }
}
