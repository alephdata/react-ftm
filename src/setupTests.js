// setup file
import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IntlProvider } from 'react-intl';
import { mount, shallow } from 'enzyme';

configure({ adapter: new Adapter() });

const messages = require('./translations/translations.json'); // en-US.json
const defaultLocale = 'en';
const locale = defaultLocale;
const filtered_messages = {};
filtered_messages[locale] = messages[locale];

global.intl = (component) => {
    return (
        <IntlProvider
            locale={locale}
            defaultLocale={defaultLocale}
            messages={filtered_messages}
        >
            {React.cloneElement(component)}
        </IntlProvider>
    );
}