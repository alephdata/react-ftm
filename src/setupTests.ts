// setup file
import '@testing-library/jest-dom';
import { createIntl } from 'react-intl';
import messages from './translations/translations.json';

export const intl = createIntl({
  locale: 'en',
  messages: messages.en,
});
