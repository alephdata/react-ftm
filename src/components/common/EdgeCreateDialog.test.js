import React from 'react';
import { EdgeCreateDialog } from './EdgeCreateDialog';
import { shallowWithIntl } from '../../setupTests';

jest.mock('types/EdgeType');

describe('<EdgeCreateDialog />', () => {
    it('Should render correctly with no properties', () => {
        const thing = {
            model: "yay"
        }
        const wrapper = shallowWithIntl(<EdgeCreateDialog entityManager={thing} />);
        expect(wrapper).toMatchSnapshot();
    });
});
