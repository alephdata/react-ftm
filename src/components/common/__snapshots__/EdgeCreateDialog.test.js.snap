// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EdgeCreateDialog /> Should render correctly with no properties 1`] = `
<Dialog
  className="large"
  icon="new-link"
  isProcessing={false}
  onClose={[Function]}
  title="Add link"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="bp3-dialog-body"
    >
      <div
        style={
          Object {
            "display": "flex",
            "flex": 1,
            "flexFlow": "row",
          }
        }
      >
        <div
          style={
            Object {
              "flexBasis": "auto",
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingRight": "1em",
            }
          }
        >
          <Blueprint3.FormGroup
            label="Source"
          >
            <Component
              allowMultiple={false}
              entitySuggestions={Array []}
              isFetching={false}
              onQueryChange={[Function]}
              onSubmit={[Function]}
              values={Array []}
            />
          </Blueprint3.FormGroup>
        </div>
        <div
          style={
            Object {
              "flexBasis": "auto",
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingRight": "1em",
            }
          }
        >
          <Blueprint3.FormGroup
            label="Type"
          >
            <EdgeTypeSelect
              items={Array []}
              onChange={[Function]}
              placeholder="Select type"
            />
          </Blueprint3.FormGroup>
        </div>
        <div
          style={
            Object {
              "flexBasis": "auto",
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingRight": "1em",
            }
          }
        >
          <Blueprint3.FormGroup
            label="Target"
          >
            <Component
              allowMultiple={false}
              entitySuggestions={Array []}
              isFetching={false}
              onQueryChange={[Function]}
              onSubmit={[Function]}
              values={Array []}
            />
          </Blueprint3.FormGroup>
        </div>
        <div
          style={
            Object {
              "flexBasis": "1%",
              "flexGrow": 0,
              "flexShrink": 1,
            }
          }
        >
          <Blueprint3.FormGroup
            label="Â "
          >
            <Blueprint3.Button
              disabled={true}
              icon="exchange"
              onClick={[Function]}
            />
          </Blueprint3.FormGroup>
        </div>
      </div>
    </div>
    <div
      className="bp3-dialog-footer"
    >
      <div
        className="bp3-dialog-footer-actions"
      >
        <Blueprint3.Button
          disabled={true}
          intent="primary"
          text="Create"
          type="submit"
        />
      </div>
    </div>
  </form>
</Dialog>
`;
